kernel/sched.c

```diff
@@ -255,9 +255,10 @@ static void reset_time_slice(void)
 	 * slice count, as we'll see those "expired" ticks arrive in a
 	 * FUTURE z_time_slice() call.
 	 */
-	_current_cpu->slice_ticks = slice_time + z_clock_elapsed();
-
-	z_set_timeout_expiry(slice_time, false);
+	if (slice_time != 0) {
```

Comment:
We can't, because the time slicing is specific to a single CPU in a SMP system and not global like a timeout.

Reply:
What about per-cpu slicing timeout?

Reply:
What about per-cpu slicing timeout?

Reply:
Could you expand more on your question?  I don't understand what you're asking for/about.

Reply:
At the moment the thread time slicing and timers are both using the same hardware providing system clock. So for example, when you are setting the timeout you have to check both slicing and timeout list to check when next timer interrupt will happen. My question is if we could implement slicing as another user of timeout list instead of bolting it on a side.

Reply:
Thanks.  So it's like the inverse of what I did in k_alarm, where the system clock was just another alarm.

My view: (a) this might be worth pursuing, but (b) it doesn't belong in and should not block this PR.  I'd put it into a new issue where it could be discussed properly.

Reply:
Yes, this does not belong here (this is why this thread starts with "Side question"). I just wanted to hear Andy's opinion.

Reply:
Using a separate counter and interrupt for timeslicing would be simpler for sure, but not all hardware has one.  And obviously seizing a device for timeslicing means that it won't be available for an app or driver to use.

Right now the architecture is that the kernel wants a single timer driver and it multiplexes it internally for use by timeouts and time slicing.  I could see value in a scheme where the kernel had separate "drivers" for those two tasks and the driver layer would do the multiplexing on hardware where they were the same underlying device.

Reply:
I am not proposing using separate timer. I just would like to setup slicing as regular entry in timeout list. If the thread will be swapped out, such entry will be removed (and readded of the next thread should be sliced too). Otherwise the kernel will be notified through callback about the need of slicing.

On the SMP this might be a bit tricky, as we will have to use IPI to forward the slicing request to the correct CPU as slicing is per-cpu.

BTW: @andyross How slicing is currently implemented on SMP targets? Are you using separate timers with interrupts bound to given CPU?